akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala

```diff
@@ -148,6 +207,21 @@ final class ORSet[A] private[akka] (
 
   type T = ORSet[A]
 
+  // Optimization for add/remove followed by merge, which is very common
+  // and merge should just fast forward to the new set
+  private var ancestor: ORSet[A] = null
```

Comment:
Hmm. This is tricky here. So this is a cache field, which is mutable. As far as I see it is safe to share this over threads because it is safe to see a null here (it is just a cache miss). What I was a bit more worried about is the check-and-act code like:

    if (this.ancestor eq null) this else this.ancestor

Although this seems to be safe since if this.ancestor becomes null after the condition have been checked it only results in assigning a null (i.e. just losing this optimization). 

However, this needed quite a bit of thought to see, so I would appreciate this to be documented here. Especially since someone in the future might do an actual "if not-null then dereference" which is no longer thread safe.